open_spiel/python/mfg/examples/mfg_Proximal_policy_optimization_pytorch.py

```diff
@@ -0,0 +1,448 @@
+        probs /= probs.sum(axis=0)
+        
+        # returns a dictionary with actions as keys and their probabilities as values
+        return {action:probs[legal_actions.index(action)] for action in legal_actions}
+
+
+def rollout(env, iter_agent, eps_agent, num_epsiodes, steps, device):
+    # generates num_epsiodes rollouts
+    info_state = torch.zeros((steps,iter_agent.info_state_size), device=device)
+    actions = torch.zeros((steps,), device=device)
+    logprobs = torch.zeros((steps,), device=device)
+    rewards = torch.zeros((steps,), device=device)
+    dones = torch.zeros((steps,), device=device)
+    values = torch.zeros((steps,), device=device)
+    entropies = torch.zeros((steps,), device=device)
+    t_actions = torch.zeros((steps,), device=device)
+    t_logprobs = torch.zeros((steps,), device=device)
+
+    step = 0
+    for _ in range(num_epsiodes):
+        time_step = env.reset()
+        while not time_step.last():
+            obs = time_step.observations["info_state"][0]
+            obs = torch.Tensor(obs).to(device)
+            info_state[step] = obs
+            with torch.no_grad():
+                t_action, t_logprob, _, _ = iter_agent.get_action_and_value(obs)
+                action, logprob, entropy, value = eps_agent.get_action_and_value(obs)
+
+            time_step = env.step([action.item()])
+
+            # iteration policy data
+            t_logprobs[step] = t_logprob
+            t_actions[step] = t_action
+
+            # episode policy data
+            logprobs[step] = logprob
+            dones[step] = time_step.last()
+            entropies[step] = entropy
+            values[step] = value
+            actions[step] = action
+            rewards[step] = torch.Tensor(time_step.rewards).to(device)
+            step += 1
+
+    return info_state, actions, logprobs, rewards, dones, values, entropies,t_actions,t_logprobs 
+
+def cal_Adv(gamma, norm, rewards,values, dones):
+    # function used to calculate the Generalized Advantage estimate
+    # using the exact method in stable-baseline3
+    with torch.no_grad():
+        next_done = dones[-1]
+        next_value = values[-1] 
+        steps = len(values)
+        returns = torch.zeros_like(rewards).to(device)
+        for t in reversed(range(steps)):
+            if t == steps - 1:
+                nextnonterminal = 1.0 - next_done
+                next_return = next_value
+            else:
+                nextnonterminal = 1.0 - dones[t + 1]
+                next_return = returns[t + 1]
+            returns[t] = rewards[t] + gamma * nextnonterminal * next_return
+
+        advantages = returns - values
+
+    if norm:
+        advantages = (advantages - advantages.mean()) / (advantages.std() + 1e-8)
+
+    return advantages, returns
+
+
+def update(update_epochs, num_minibatch, obs, logprobs, actions, advantages, returns, t_actions, t_logprobs, optimizer_actor, optimize_critic, agent, alpha = 0.5, t_eps = 0.2, eps = 0.2):
```

Comment:
All lines should be under 80 characters.

Probably best to take a quick read of the Google Python style guide: https://google.github.io/styleguide/pyguide.html

Reply:
Code checked with pylint, now lines are under 80 characters, extra spaces were removed and the number of arguments / function size reduced for better readability.